---
title: "Bovreproseq Results Report"
author: "Generated using [Bovreproseq](https://github.com/dhineshp565/Bovreproseq) pipeline"
date: "`r Sys.Date()`"
output: 
  html_document:
  output_file: "Bovreposeq_report_`r Sys.Date()`.html"
params:
  csv: ""
  krona: ""
---

```{r setup, include=FALSE}

library(knitr)
library(base64enc)
library(htmltools)
library(kableExtra)
```


<br>
```{css, echo=FALSE}
.table caption {
    color: darkblue;
    font-weight: bold;
}
```

### Click dropdown to select sample {.tabset .tabset-dropdown}
```{r, results='asis',echo=FALSE}

samplelist <- read.csv(params$csv,header=FALSE,sep = ',')[-1,]
sampleid <- samplelist[,c(1)]

for (i in sampleid){
  cat("####", i, "\n")
  
  results <- (paste(i,"_results.csv",sep=""))
  results_df <- read.csv(results,header = TRUE,sep = ",")
  colnames (results_df) <- c("SAMPLEID","PATHOGEN","RESULT","QC STATUS") # column names
  #print (knitr::kable(results_df,align="llll",caption = "RESULTS SUMMARY - Based on abricate results")) 
  # Create a logical vector indicating where the "Positive" values are
  # Assuming results_df is your data frame and RESULT is the column containing "Positive"
  positive_rows <- results_df$RESULT == "Positive"
  
  internal_control_negative_rows <- results_df$PATHOGEN == "Internal amplification control" & results_df$RESULT == "Negative"


  # Generate the table with kable
  table_html <- knitr::kable(results_df, align = "llll", caption = "RESULTS SUMMARY - Based on abricate results with custom Bovreproseq database. For more info: Seemann T, Abricate, Github https://github.com/tseemann/abricate") %>%
    kable_styling()

  # Apply row highlighting
  if (any(positive_rows)) {
    table_html <- table_html %>%
      kable_styling(full_width = TRUE) %>%
      row_spec(which(positive_rows), background = "lightgreen")
  }
  if (any(internal_control_negative_rows)) {
    table_html <- table_html %>%
      kable_styling(full_width = TRUE) %>%
      row_spec(which(internal_control_negative_rows), background = "orange")
  }

  # Print the HTML table
  print(table_html)

  cat('\n\n<!-- -->\n\n')
  cat('\n\n<!-- -->\n\n')

  #use mappedreads.txt and display column 1 (amplicon name) and 3 (no. of mapped reads)
  mapped_reads <- (paste(i,"_mappedreads.txt",sep="")) # filename
  df <- read.csv(mapped_reads,header = TRUE,sep=" ") # dataframe
  mapped_stat<-df[,c(1,3)] # sliced dataframe
  colnames (mapped_stat) <- c("AMPLICON","MAPPED_READS") # column names
  print(knitr::kable(mapped_stat,align="ll",caption = "NO. OF MAPPED READS. Generated using minimap2 (Li, H. (2018).Bioinformatics, 34:3094-3100. https://doi.org/10.1093/bioinformatics/bty191) and samtools (https://doi.org/10.1093/gigascience/giab008)") %>%
    kable_styling()) # displat table
  
  cat('\n\n<!-- -->\n\n') 
 
  # display abricate results
  abricate_ids <- (paste(i,"_abricate.csv",sep=""))
  abricate_raw <- read.csv(abricate_ids,header = TRUE,sep = "\t")
  abricate_final <- abricate_raw[,c(2,6,10,11,12,15)]
  colnames(abricate_final) <- c("SEQUENCE","GENE","%COVERAGE","%IDENTITY","DATABASE","REFERENCE") # nolint
  print(knitr::kable(abricate_final,align = "llcccl",caption = "Consensus compared to reference database. Generated using abricate (Abricate, Github https://github.com/tseemann/abricate) and custom database") %>%
    kable_styling()) # nolint: line_length_linter.
  
   cat('\n\n<!-- -->\n\n')

  mlst_ids <- (paste(i,"_MLST_results.csv",sep=""))
  mlst_raw <- read.csv(mlst_ids,header = TRUE,sep = "\t")
  mlst_final <- mlst_raw[,c(1,3,11)]

  if (grepl("fetus", mlst_final$ORGANISM)) {
  print(knitr::kable(mlst_final,align = "lcl",caption = "MLST tool from github (https://github.com/tseemann/mlst). Scans contig files against traditional PubMLST typing schemes (https://pubmlst.org/) ") %>%
    kable_styling())
  }
  cat('\n\n<!-- -->\n\n')

  # Display fasta sequence
  fasta_id <- (paste(i,"_medaka_consensus.fasta",sep=""))
  fasta_file <- readLines(fasta_id)
  # Display the content
  cat("```\n")
  cat(fasta_file, sep = "\n")
  cat("\n```\n")
  # Encode the file content as base64
  encoded_content <- base64encode(charToRaw(paste(fasta_file, collapse = "\n")))
  # Get file name from file path and remove extra underscores
  file_name <- gsub("_+", "_", basename(fasta_id))
  # Provide a download link for the encoded FASTA content with file name
  cat('<a href="data:text/plain;charset=utf-8;base64,', encoded_content, '" download="', file_name, '">Download FASTA file</a>')
  
  cat('\n\n<!-- -->\n\n')


}
```
<br>

<br>

## Taxonomic classification of raw reads   
##### Generated using kraken2 (kraken2 (https://ccb.jhu.edu/software/kraken2/) and krona (https://github.com/marbl/Krona)
```{r, echo=FALSE}
library("htmltools")
htmltools::tags$iframe(
  src = base64enc::dataURI(file= params$krona, mime="text/html; charset=UTF-8"),
  style="border:10; position:relative; top:0; left:; right::; bottom:; width:100%; height:800px"
)
```